### PR TITLE
Use let const as blockscoped variables

### PR DESCRIPTION
**Pull request by timse**
**What kind of change does this PR introduce?**

refactoring

**Did you add tests for your changes?**

N/A

**If relevant, link to documentation update:**

N/A

**Summary**

During the refactoring a lot of the `var` assignments where translated to `let`, `const` without taking into account that they are blockscoped.
On top of that `let` should only be used when a reference is reassigned in the code.
This allows for easier reading as something defined as a `const` can be considered "never reassigned" while something explicitly labeled as `let` will be reassigned later on.

**Does this PR introduce a breaking change?**

No

**Other information**

There are a couple of interesting finds that i stumpled upon while doing this change:
- in https://github.com/timse/webpack/commit/c3abf8f5cc6103480f04b101f6b80d081b232c0e some variables are assigned and checked and reassigne but are always false
- was this console.log intended? https://github.com/timse/webpack/commit/bd386ecf06c0886504fe2e957fc108b2c0b96b3a
- one assignment here will always be overridden, was that intended? https://github.com/timse/webpack/commit/424a0aead9fdf53f2eb5a0d1c72a42bcdbe2278b

### CODE REVIEW

File: lib/FlagDependencyExportsPlugin.js [hunk 1]
```diff
@@ -45,8 +45,7 @@ class FlagDependencyExportsPlugin {
 					if(exportDeps) {
 						exportDeps.forEach((dep) => {
 							const depIdent = dep.identifier();
-							let array = dependencies[depIdent];
-							if(!array) array = dependencies[depIdent] = [];
+							const array = dependencies[depIdent] = dependencies[depIdent] || [];
```
Comment (wtgtybhertgeghgtwtg) [inline lib/FlagDependencyExportsPlugin.js]: Wouldn't it be cleaner as
```
if (!dependencies[depIdent]) {
  dependencies[depIdent] = [];
}
const array = dependencies[depIdent];
```

Comment (timse) [inline lib/FlagDependencyExportsPlugin.js]: I like that :)

Comment (timse) [inline lib/FlagDependencyExportsPlugin.js]: hm actually how about this :P?

```
const depIdent = dep.identifier();
const array = dependencies[depIdent];
// if this was not yet initialized
// initialize it as an array containing the module and stop
if(!array) {
  dependencies[depIdent] = [module];
  return;
}

// check if this module is known
// if not, add it to the dependencies for this identifier
if(array.indexOf(module) < 0)
  array.push(module);
```

Comment (wtgtybhertgeghgtwtg) [inline lib/FlagDependencyExportsPlugin.js]: You could change the return and second if to an else if like
```
const depIdent = dep.identifier();
const array = dependencies[depIdent];
if (!array) {
  dependencies[depIdent] = [module]
} else if (array.indexOf(module) < 0) {
  array.push(module);
}
```
but either way works.

Comment (timse) [inline lib/FlagDependencyExportsPlugin.js]: not a big fan of `else ifs` ;P

**Comment by Tom910** [comment]
Eslint have suitable rule prefer-const - http://eslint.org/docs/rules/prefer-const

**Comment by timse** [comment]
i think i need a little help for the statsCases test order, when is there a case where the asset contains a "!"?

**Comment by sokra** [comment]
i. e. `stats: { modulesSort: "!id" }`. It sorts by id but in inverse order.

File: lib/Compilation.js [hunk 2]
```diff
@@ -1146,18 +1144,14 @@ class Compilation extends Tapable {
...
-			this.assets[file] = module.assets[name];
-			this.applyPlugins2("module-asset", module, file);
-		}
-
 		for(let i = 0; i < this.modules.length; i++) {
-			module = this.modules[i];
+			const module = this.modules[i];
 			if(module.assets) {
-				Object.keys(module.assets).forEach(cacheAssetsAndApplyPlugins, this);
+				Object.keys(module.assets).forEach((assetName) => {
```
Comment (sokra) [inline lib/Compilation.js]: creating functions in loops in expensive. Move it above the loop.

File: lib/FlagDependencyExportsPlugin.js [hunk 3]
```diff
@@ -45,10 +45,17 @@ class FlagDependencyExportsPlugin {
 					if(exportDeps) {
 						exportDeps.forEach((dep) => {
 							const depIdent = dep.identifier();
-							let array = dependencies[depIdent];
-							if(!array) array = dependencies[depIdent] = [];
-							if(array.indexOf(module) < 0)
-								array.push(module);
+							// if this was not yet initialized
+							// initialize it as an array containing the module and stop
+							if(!dependencies[depIdent]) {
```
Comment (sokra) [inline lib/FlagDependencyExportsPlugin.js]: The array should be stored in a variable for performance reasons.

File: test/configCases/plugins/define-plugin/index.js [hunk 4]
```diff
@@ -1,89 +1,104 @@
-it("should define FALSE", function() {
-	FALSE.should.be.eql(false);
-	(typeof TRUE).should.be.eql("boolean");
-	var x = require(FALSE ? "fail" : "./a");
-	var y = FALSE ? require("fail") : require("./a");
-});
+/* globals define, it, should */
+define("DefinePlugin", function() {
```
Comment (sokra) [inline test/configCases/plugins/define-plugin/index.js]: It' describe actually, but not needed here, as the folder name is already used as suited name.

File: test/configCases/plugins/define-plugin/index.js [hunk 5]
```diff
@@ -1,89 +1,104 @@
...
+		f // <- no semicolon here
+		OBJECT;
+	});
+
+	it("should not explode on recursive typeof calls", function() {
+		(typeof wurst).should.eql("undefined"); // <- is recursivly defined in config
+	});
+
+	it("should not explode on recursive statements", function() {
+		try {
```
Comment (sokra) [inline test/configCases/plugins/define-plugin/index.js]:
``` js
(function() {

}).should.throw("suppe is not defined");
```

File: lib/Compilation.js (hunk 2 above)
Comment (sokra) [inline lib/Compilation.js]: Ok doesn't work here.

**Comment by sokra** [comment]
Thanks for the great work. I see that there happened a lot weird stuff during ES6 migration.

File: lib/Compilation.js (hunk 2 above)
Comment (timse) [inline lib/Compilation.js]: let me try something

**Comment by sokra** [comment]
Thanks

**Comment by timse** [comment]
thanks for merging :)

**Pull request merged**